plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerDelegate.java

```diff
@@ -1,37 +1,99 @@
 package com.google.cloud.tools.eclipse.appengine.localserver.server;
 
+import java.util.ArrayList;
+import java.util.List;
+
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Status;
+import org.eclipse.jst.server.core.IWebModule;
 import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IModuleType;
+import org.eclipse.wst.server.core.internal.facets.FacetUtil;
 import org.eclipse.wst.server.core.model.ServerDelegate;
 
+@SuppressWarnings("restriction") // For FacetUtil
 public class LocalAppEngineServerDelegate extends ServerDelegate {
+  private static final String SERVLET_MODULE_FACET = "jst.web";
+  private static final String ATTR_APP_ENGINE_SERVER_MODULES = "app-engine-server-modules-list";
 
+  /**
+   * Returns OK status if the projects associated with modules to be added support the App Engine
+   * runtime, otherwise returns an ERROR status.
+   */
   @Override
   public IStatus canModifyModules(IModule[] add, IModule[] remove) {
-    // todo what do we test here?
+    if (add != null) {
+      for (IModule module : add) {
+        if (module.getProject() != null) {
+          IStatus status = FacetUtil.verifyFacets(module.getProject(), getServer());
+          if (status != null && !status.isOK()) {
+            return status;
+          }
+        }
+      }
+    }
     return Status.OK_STATUS;
   }
 
+  /**
+   * If the module is a web module returns the utility modules contained within its
+   * WAR, otherwise returns an empty list.
+   */
   @Override
   public IModule[] getChildModules(IModule[] module) {
-    // TODO Auto-generated method stub
-    return null;
+    if ((module != null) && (module[0] != null) && (module[0].getModuleType() != null)) {
+      IModule thisModule = module[module.length - 1];
```

Comment:
The `if` on the previous line checks for the item in the 0th position and this one works with the last item in the array, is this consistent?